### PR TITLE
[FIX] base: prevent parallel module installation

### PR DESCRIPTION
**Pull request by yelizariev**
Parallel module installation is a quite bad idea and had to be forbidden.

Before this commit user may start module installation, give up with waiting
loading page, try to start the installation again. It may lead to different
error like SERIALIZATION_FAILURE, Failed to load registry, etc.

At the end it can make db inconsistent. For example, we found DB [1] that has
`sale`, `stock`, `account` modules installed, but not `sale_account`,
`sale_stock`, which are supposed to be installed automatically.

Fix it by checking modules with `to %` state.

It's possible to have module states equal `to install`, while no module
operation is running. However, this scenario is very rare [2] and administrator can
always fix it manually by using "Cancel install" button at the module form.

[1]: https://online.sentry.io/issues/4041525003/

[2]

It case if the process is failed, the modules states are
reset by method `reset_modules_state` from `odoo/modules/loading.py`. The method
is called right after the failure or by cron after 5 hours of waiting in a cron
queue.

https://github.com/odoo/odoo/blob/0fab57e4eb6637d10e55f473d405b514b8becd0b/odoo/modules/registry.py#L89-L93

https://github.com/odoo/odoo/blob/17cd7fcd6d5334be884176c36731d1662d543960/odoo/addons/base/models/ir_cron.py#L177-L181

### CODE REVIEW

**Comment by rvalyi** [comment]
hello, I'm not sure if it is impacted or not. But may be it would be nice if a module update (not installation) could still be performed in parallel as it could speed up database migrations quite a bit if the migration tool is able to launch several migration workers. If a module upgrade triggers a new installation may be it would clash with this change.

**Comment by yelizariev** [comment]
@xmo-odoo 
In short, we both agree that concurrency problem exist.

What about following solution

1. Move the lock to `load_modules` function since it's the place where all module operations are actually processed
2. Change lock query to `LOCK TABLE ir_cron IN EXCLUSIVE MODE NOWAIT`, so it will work even if there is no cron records in the able.

**Comment by HydrionBurst** [comment]
> At the end it can make db inconsistent. For example, we found DB [1](https://online.sentry.io/issues/4041525003/) that has `sale`, `stock`, `account` modules installed, but not `sale_account`, `sale_stock`, which are supposed to be installed automatically.

Follow this one. I think the result is acceptable. Unlike `upgrade`, `install` doesn't have `-end` phase. The database is relatively consistent with registry when error raised during installation. The user only have to check if the APP is installed after the error dialog and if not install them again.

File: odoo/addons/base/models/ir_module.py
Comment (yelizariev) [reply odoo/addons/base/models/ir_module.py:592]: Thanks for review.

I've updated the PR to acquire lock before module operations that have commit after processing the module. What do you think?

Comment (yelizariev) [reply odoo/addons/base/models/ir_module.py]: You are right again. The lock, cr.commit and cron mixed together make it hard to add extra locks to fix the problem.

What about simple checking of modules states? If there are `to install` modules, then we raise `UserError`. If something fails, then `reset_modules_state` is our best friend (and *Cancel install* button at the module form)

(I've updated the patch and PR description)

**Comment by HydrionBurst** [comment]
I think it is good enough for the problem we want to solve.
I suggest 
1. we change the name `transient` to another term e.g. `updating` which means these modules are being udpated by another request. 
2. your solution is only for `_button_immediate_function` which means when it is called, the registry should be ready. So we don't need to create a new function `get_transient_modules`, a simple `search_count` should solve the problem.
3. move the code together with `self._cr.execute("SELECT * FROM ir_cron FOR UPDATE NOWAIT")` since they will raise the same error

**Comment by HydrionBurst** [comment]
Another suggestion:
Now, since `to xxx` state will block the install/upgrade. Maybe we can provider a filter for the search bar to list all updating modules.
1. the user will know when modules are installable/upgradable
2. if some `state`s go wrong (I don't find the corner case) and become 'to xxx' forever, it would be easier for users to fix the problem from UI by cancelling install/upgrade

**Comment by HydrionBurst** [comment]
I think the idea is good enough for the problem we want to solve.
I suggest the following change to group up all tricky locks

```diff
         try:
-            # This is done because the installation/uninstallation/upgrade can modify a currently
-            # running cron job and prevent it from finishing, and since the ir_cron table is locked
-            # during execution, the lock won't be released until timeout.
+            # This is done because
+            # 1. the installation/uninstallation/upgrade can modify a currently running cron job
+            # 2. another request is installing/uninstalling/upgrading modules.
+            # and prevent it from finishing
             self._cr.execute("SELECT * FROM ir_cron FOR UPDATE NOWAIT")
-        except psycopg2.OperationalError:
-            raise UserError(_("Odoo is currently processing a scheduled action.\n"
+            assert not self.search_count([('state', 'in', ('to install', 'to upgrade', 'to remove'))], limit=1)
+        except (psycopg2.OperationalError, AssertionError):
+            raise UserError(_("Odoo is currently processing a scheduled action or updating.\n"
                               "Module operations are not possible at this time, "
                               "please try again later or contact your system administrator."))
         function(self)

         self._cr.commit()
```

1. I change the comment because: 
   the original comment for the `self._cr.execute("SELECT * FROM ir_cron FOR UPDATE NOWAIT")` was not accurate, the lock will be released when `self._cr.commit()` but not `ir_cron table is locked during execution`.
2. I only use one try except because:
    we need the cooperation of `SELECT FOR UPDATE` and `search_count` to prevent paralleled installation/upgrade. And we cannot tell if the `psycopg2.OperationalError` is raised because of `cron` or another install/upgrade request.  The risky concurrent window is only the post processing time when all `to xxx` modules have been marked `installed`/`uninstalled` which is quite small and usually doesn't cause problems
3. I don't show the `to xxx` modules to users becasue:
     they can use the search bar to get all these updating modules with a filter

cc @rco-odoo

**Comment by yelizariev** [comment]
@HydrionBurst I applied your patch with some modifications in comments and error message. Feel free  to propose the final version

**Comment by yelizariev** [comment]
@HydrionBurst done

Please check my updated code comments and commit description. I know you guys take such things seriously. You can update it directly or post a review. Thanks.

**Comment by HydrionBurst** [comment]
LGTM
I think the fix is elegant, but I want to wait for some review from @rco-odoo

**Comment by yelizariev** [comment]
Also, https://online.sentry.io/issues/3937959802

**Review by Julien00859 — APPROVED** [review]
better than nothing but there is a critical zone in between `search_count` and `execute(... FOR UPDATE)`, in case one worker passes the first check and get context-switched just before attempting to acquire the lock, another worker can `commit()` its own installation.

not great, not terrible and we couldn't come with a better way with @HydrionBurst :/

@robodoo r+

**Comment by yelizariev** [comment]
Interesting :thinking:
I'll check 
image

**Comment by HydrionBurst** [comment]
I don't understand
> As a workaround we could ignore search_count check.

Standalone tests are not transactional, it is no a huge surprise that the database might be contaminated by another test.
But I think it is always better forcely fixing the wrong data instead of ingoring some process
Maybe you can `reset_modules_state` (+ `commit` optional) and then `button_immediate_install`

File: addons/account/tests/test_account_all_l10n.py
Comment (yelizariev) [inline addons/account/tests/test_account_all_l10n.py:21]: Try r+ again?

Comment (HydrionBurst) [reply addons/account/tests/test_account_all_l10n.py:21]: Could you check which standalone test contaminate the db? It makes more sence to me to `reset_module_state` in that test.

Comment (Julien00859) [reply addons/account/tests/test_account_all_l10n.py:21]: @yelizariev can you elaborate here and in the commit message why is resetting the module states required inside that test? My feeling is that you added this function call as an attempt to make the test passes but without real understanding why some modules were to-install/to-remove/to-update and without challenging whether resetting the states was correct.

Comment (yelizariev) [reply addons/account/tests/test_account_all_l10n.py:21]: I really don't know what is going on runbot with those standalone tests. I was thinking that you guys have some ideas,

The build starts Odoo and runs `test_all_l10n` and fails immediately.

https://runbot.odoo.com/runbot/build/49846664
http://runbot137.odoo.com/runbot/static/build/49846664-master/logs/l10n_standalone.txt

May I get some insights how those build work? It gets DB dump from another build, right? Which build?

Also, I tried to restore following DB locally

image

but modules states are fine

image

File: odoo/addons/base/models/ir_module.py
Comment (HydrionBurst) [inline odoo/addons/base/models/ir_module.py:556]: OK, I am a blind reviewer 😭 
```suggestion
        if self.search_count([('state', 'in', ('to install', 'to upgrade', 'to remove'))], limit=1):
```

Comment (yelizariev) [reply odoo/addons/base/models/ir_module.py]: :grimacing:

**Review by Julien00859 — CHANGES_REQUESTED** [review]
I decided to test the problem on master and on this branch, the result was the same: I was able to install several modules at the same time.

On a fresh db, I was able to install CRM and Sales at the same time in two tabs. None of the tab got rejected with a UserError. At the end of the installation, both tab showed an Internal Server Error and there were serialization issues reported in the logs. I was able to access /web on both tabs and everything *seems* to be working but nothing prevented "parallel module installation".

This is in track with what I said earlier, there are critical zones and it is much easier to end up in that critical zone than I thought.

I think that doing two queries, one to check the states and one to acquire the log just *cannot* produce any viable result. We'll always end up with a critical zone and I am now absolutely confident that it is *trivial* to end up in that critical zone. We have to *both* acquire the lock *and* be sure that a previous installation is not still on-going (but lost the lock because it commited mid-request) *at the same time*. If we keep on checking for to-update/to-install modules then we also have to make sure we correctly detect changes done in another transaction even with our repeatable-read mode of isolation (don't change the isolation level!).

**Pull request closed without merging**
